core/src/main/scala/org/broadinstitute/dsde/rawls/workspace/MultiCloudWorkspaceService.scala

```diff
@@ -216,21 +229,86 @@ class MultiCloudWorkspaceService(override val ctx: RawlsRequestContext,
     } yield clone
 
   def cloneAzureWorkspace(sourceWorkspace: Workspace,
-                          project: RawlsBillingProject,
                           profile: ProfileModel,
-                          request: WorkspaceRequest
-  ): Future[Workspace] = Future.failed(
-    RawlsExceptionWithErrorReport(
-      ErrorReport(StatusCodes.NotImplemented, "WOR-625")
-    )
-  )
+                          request: WorkspaceRequest,
+                          parentContext: RawlsRequestContext
+  ): Future[Workspace] = {
+
+    def createNewWorkspaceRecord(): Future[Workspace] =
+      dataSource.inTransaction { access =>
+        for {
+          _ <- failIfWorkspaceExists(request.toWorkspaceName)
+          workspaceId = UUID.randomUUID()
+          newWorkspace <- createMultiCloudWorkspaceInDatabase(
+            workspaceId.toString,
+            request.toWorkspaceName,
+            request.attributes,
+            access,
+            parentContext
+          )
+        } yield newWorkspace
+      }
+
+    for {
+      // Guard for forwards compatibility:
+      // We only support cloning azure workspace storage containers into the Azure Config's
+      // `defaultRegion` for now.
+      _ <- request.bucketLocation.traverse_(throwUnlessIsDefaultRegion)
+
+      // The call to WSM is asynchronous. Before we fire it off, allocate a new workspace record
```

Comment:
Can't we wait to write the record until we at least get back a success from starting the clone? Similar to what we do on workspace creation?

Reply:
I think we should change that behaviour too to avoid contention issues - reserve the workspace id before using it.

Reply:
I still don't understand why we don't wait to verify that kicking off the clone succeeded (by checking `cloneResult`, which I would probably rename as `cloneJobResult`) before creating the Rawls record. That operation is synchronous.

Reply:
If a user sends two requests in quick succession (script error, bug in UI, whatever, if it could happen it will happen), we could send two clone requests for the same workspace name. One will then subsequently fail with a resource conflict. The idea is I want to allocate/reserve the workspace name and then send the clone request so that this sort of thing doesn't happen.